model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java

```diff
@@ -459,25 +459,32 @@ public KeycloakSession getKeycloakSession() {
     @Override
     public RealmModel getRealmByName(String name) {
         String cacheKey = getRealmByNameCacheKey(name);
+        if (invalidations.contains(cacheKey)) {
+            return getRealmDelegate().getRealmByName(name);
+        }
```

Comment:
@ahus1 here we will have multiple threads pounding the database, even before we get to the `synchronized` part.

Suppose the following conditions:
1. Realm `myRealm` has just been invalidated and thus `invalidations.contains(cacheKey)` is `true`

Now suppose _n_ threads service requests that come in at the same time.
This means line 462 will be evaluated _n_ times to `true` concurrently.  Subsequently line 463 will be invoked _n_ times concurrently.  

`getRealmDelegate().getRealmByName(name)` on 463 will invoke DB I/O.  This will be done _n_ times concurrently and thus cause the stampede phenomena.

Reply:
@JesseEstum - some background on `KeycloakSession`: It captures the transaction context of a single request, and starts together with the request and ends when the request ends. It can be nested to support nested transactions, but most of the time it is not. 

The `RealmCacheSession` is instantiated during a request and bound to an instance of `KeycloakSession`, so it has the same lifetime (usually even shorter). 

The  attributes `invalidations` and `listInvalidations` are attributes local to `RealmCacheSession`, and only contain information from the current request. Once there are local changes to a realm, they are filled with the invalidations which will be published once the current request completes. 

Due to this, there can't be a cache stampede due to them, as there are no concurrent requests on them.

Reply:
@ahus1 Thanks for the background/context.  I am admittedly not familiar with the overall request/response and object model in Keycloack so this is useful to understand.  With this understanding now in hand, I agree my [previous comment](https://github.com/keycloak/keycloak/pull/22989#discussion_r1322018201) is not correct.